plugins/dbus_announce.c

```diff
@@ -0,0 +1,143 @@
+#include "system.h"
+
+#include <dbus/dbus.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <unistd.h>
+#include <rpm/rpmlog.h>
+#include <rpm/rpmts.h>
+#include <rpm/rpmdb.h>
+#include "lib/rpmplugin.h"
+
+struct dbus_announce_data {
+    DBusConnection * bus;
+    int logging;
+    int old_dboffset;
+};
+
+static rpmRC dbus_announce_init(rpmPlugin plugin, rpmts ts)
+{
+    struct dbus_announce_data * state = rcalloc(1, sizeof(*state));
+    rpmPluginSetData(plugin, state);
+    return RPMRC_OK;
+}
+
+static rpmRC open_dbus(rpmPlugin plugin, rpmts ts)
+{
+    struct stat st;
+    DBusError err;
+    int rc = 0;
+    struct dbus_announce_data * state = rpmPluginGetData(plugin);
+
+    /* Assume we are logging but... */
+    state->logging = 1;
+
+    /* ...don't log test transactions */
+    if (rpmtsFlags(ts) & (RPMTRANS_FLAG_TEST|RPMTRANS_FLAG_BUILD_PROBS))
+	state->logging = 0;
+
+    /* ...don't log chroot transactions */
+    if (!rstreq(rpmtsRootDir(ts), "/"))
+	state->logging = 0;
+
+    /* Don't open */
+    if (!state->logging || state->bus)
+	return RPMRC_OK;
+
+    if (lstat("/run/systemd/system/", &st) == 0) {
+        if (S_ISDIR(st.st_mode)) {
+	    // initialise the error value
+	    dbus_error_init(&err);
+
+	    state->bus = dbus_bus_get_private(DBUS_BUS_SYSTEM, &err);
+	    if (dbus_error_is_set(&err)) {
+		fprintf(stderr, "Connection Error (%s)\n", err.message);
+		dbus_error_free(&err);
+	    }
+	    if (state->bus) {
+		rc = dbus_bus_request_name(state->bus, "org.rpm.announce",
+				  DBUS_NAME_FLAG_REPLACE_EXISTING , &err);
+	    }
+	    if (dbus_error_is_set(&err)) {
+		//fprintf(stderr, "Name Error (%s)\n", err.message);
+		dbus_error_free(&err);
+	    }
+	    if (DBUS_REQUEST_NAME_REPLY_PRIMARY_OWNER != rc) {
+		dbus_connection_close(state->bus);
+		dbus_connection_unref(state->bus);
+		state->bus = NULL;
+		state->logging = 0;
+		return RPMRC_NOTFOUND;
+	    }
+            return RPMRC_OK;
+        }
+    }
+
+    return RPMRC_NOTFOUND;
```

Comment:
The return code logic is confusing, with multiple returns in space of just a few lines.